File: codecraft-frontend/convex/projects.ts
import { v } from "convex/values";
import { mutation, query } from "./_generated/server";
import { Id } from "./_generated/dataModel";

// Create a new project
export const createProject = mutation({
  args: {
    name: v.string(),
    description: v.optional(v.string()),
    userId: v.id("users"),
    files: v.optional(v.array(v.object({
      path: v.string(),
      content: v.string(),
      language: v.string(),
    }))),
  },
  handler: async (ctx, args) => {
    const now = Date.now();
    
    // Create the project
    const projectId = await ctx.db.insert("projects", {
      name: args.name,
      description: args.description,
      userId: args.userId,
      createdAt: now,
      updatedAt: now,
    });

    // Add files if provided
    if (args.files) {
      for (const file of args.files) {
        await ctx.db.insert("projectFiles", {
          projectId,
          path: file.path,
          content: file.content,
          language: file.language,
          createdAt: now,
          updatedAt: now,
        });
      }
    }

    return projectId;
  },
});

// Get a project by ID with files and chat messages
export const getProject = query({
  args: { projectId: v.id("projects") },
  handler: async (ctx, args) => {
    const project = await ctx.db.get(args.projectId);
    if (!project) return null;

    // Get project files
    const files = await ctx.db
      .query("projectFiles")
      .withIndex("by_project", (q) => q.eq("projectId", args.projectId))
      .collect();

    // Get chat messages
    const chatMessages = await ctx.db
      .query("chatMessages")
      .withIndex("by_project_timestamp", (q) => q.eq("projectId", args.projectId))
      .order("asc")
      .collect();

    return {
      ...project,
      files: files.map(file => ({
        path: file.path,
        content: file.content,
        language: file.language,
      })),
      chat_messages: chatMessages.map(msg => ({
        role: msg.role,
        content: msg.content,
        timestamp: new Date(msg.timestamp).toISOString(),
      })),
    };
  },
});

// Get all projects for a user
export const getUserProjects = query({
  args: { userId: v.id("users") },
  handler: async (ctx, args) => {
    const projects = await ctx.db
      .query("projects")
      .withIndex("by_user", (q) => q.eq("userId", args.userId))
      .order("desc")
      .collect();

    // Get files and chat messages for each project
    const projectsWithDetails = await Promise.all(
      projects.map(async (project) => {
        const files = await ctx.db
          .query("projectFiles")
          .withIndex("by_project", (q) => q.eq("projectId", project._id))
          .collect();

        const chatMessages = await ctx.db
          .query("chatMessages")
          .withIndex("by_project", (q) => q.eq("projectId", project._id))
          .collect();

        return {
          id: project._id,
          name: project.name,
          description: project.description,
          created_at: new Date(project.createdAt).toISOString(),
          updated_at: new Date(project.updatedAt).toISOString(),
          files: files.map(file => ({
            path: file.path,
            content: file.content,
            language: file.language,
          })),
          chat_messages: chatMessages.map(msg => ({
            role: msg.role,
            content: msg.content,
            timestamp: new Date(msg.timestamp).toISOString(),
          })),
        };
      })
    );

    return projectsWithDetails;
  },
});

// Update project files
export const updateProjectFiles = mutation({
  args: {
    projectId: v.id("projects"),
    files: v.array(v.object({
      path: v.string(),
      content: v.string(),
      language: v.string(),
    })),
  },
  handler: async (ctx, args) => {
    const now = Date.now();

    // Delete existing files
    const existingFiles = await ctx.db
      .query("projectFiles")
      .withIndex("by_project", (q) => q.eq("projectId", args.projectId))
      .collect();

    for (const file of existingFiles) {
      await ctx.db.delete(file._id);
    }

    // Insert new files
    for (const file of args.files) {
      await ctx.db.insert("projectFiles", {
        projectId: args.projectId,
        path: file.path,
        content: file.content,
        language: file.language,
        createdAt: now,
        updatedAt: now,
      });
    }

    // Update project timestamp
    await ctx.db.patch(args.projectId, {
      updatedAt: now,
    });

    return true;
  },
});

// Add a chat message
export const addChatMessage = mutation({
  args: {
    projectId: v.id("projects"),
    role: v.union(v.literal("user"), v.literal("assistant")),
    content: v.string(),
    userId: v.optional(v.id("users")),
  },
  handler: async (ctx, args) => {
    const now = Date.now();

    const messageId = await ctx.db.insert("chatMessages", {
      projectId: args.projectId,
      role: args.role,
      content: args.content,
      timestamp: now,
      userId: args.userId,
    });

    // Update project timestamp
    await ctx.db.patch(args.projectId, {
      updatedAt: now,
    });

    return messageId;
  },
});

// Delete a project
export const deleteProject = mutation({
  args: { projectId: v.id("projects") },
  handler: async (ctx, args) => {
    // Delete project files
    const files = await ctx.db
      .query("projectFiles")
      .withIndex("by_project", (q) => q.eq("projectId", args.projectId))
      .collect();

    for (const file of files) {
      await ctx.db.delete(file._id);
    }

    // Delete chat messages
    const messages = await ctx.db
      .query("chatMessages")
      .withIndex("by_project", (q) => q.eq("projectId", args.projectId))
      .collect();

    for (const message of messages) {
      await ctx.db.delete(message._id);
    }

    // Delete the project
    await ctx.db.delete(args.projectId);

    return true;
  },
});

// Generate code using AI (action that calls external API)
export const generateCode = mutation({
  args: {
    prompt: v.string(),
    userId: v.id("users"),
  },
  handler: async (ctx, args) => {
    // This will be implemented to call the backend API
    // For now, return a mock response
    const now = Date.now();
    
    // Create a new project with generated code
    const projectId = await ctx.db.insert("projects", {
      name: "Generated Project",
      description: "AI-generated project",
      userId: args.userId,
      createdAt: now,
      updatedAt: now,
    });

    // Add sample files (this will be replaced with actual AI generation)
    const sampleFiles = [
      {
        path: "/App.js",
        content: `import React from 'react';

function App() {
  return (
    <div className="App">
      <h1>Hello World!</h1>
      <p>This is a generated React app.</p>
    </div>
  );
}

export default App;`,
        language: "javascript",
      },
      {
        path: "/index.html",
        content: `<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Generated App</title>
</head>
<body>
    <div id="root"></div>
</body>
</html>`,
        language: "html",
      },
    ];

    for (const file of sampleFiles) {
      await ctx.db.insert("projectFiles", {
        projectId,
        path: file.path,
        content: file.content,
        language: file.language,
        createdAt: now,
        updatedAt: now,
      });
    }

    return {
      project_id: projectId,
      name: "Generated Project",
      description: "AI-generated project",
      files: sampleFiles,
    };
  },
});

File: codecraft-backend/endpoints/projects.py
from fastapi import APIRouter, HTTPException, Depends
from typing import List, Dict, Any
import uuid
from datetime import datetime

from models.project import (
    ProjectCreate, ProjectUpdate, ProjectResponse,
    GenerateCodeRequest, GenerateCodeResponse,
    ChatRequest, ChatResponse, ProjectFile, ChatMessage
)
from services.project_service import ProjectService

router = APIRouter(prefix="/api/projects", tags=["projects"])
project_service = ProjectService()

# In-memory storage for demo (replace with Convex DB integration)
projects_db: Dict[str, Dict[str, Any]] = {}

@router.post("/generate", response_model=GenerateCodeResponse)
async def generate_code(request: GenerateCodeRequest):
    """Generate code based on user prompt"""
    try:
        response = await project_service.generate_code(request)
        return response
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/create", response_model=ProjectResponse)
async def create_project(request: ProjectCreate):
    """Create a new project"""
    try:
        project_id = str(uuid.uuid4())
        
        # Generate initial code if prompt provided
        initial_files = {}
        initial_chat = []
        
        if request.initial_prompt:
            # Generate code based on initial prompt
            gen_request = GenerateCodeRequest(
                prompt=request.initial_prompt,
                template=request.template
            )
            gen_response = await project_service.generate_code(gen_request)
            
            # Convert generated files to ProjectFile objects
            initial_files = project_service.convert_ai_files_to_project_files(gen_response.files)
            
            # Add initial chat messages
            initial_chat = [
                ChatMessage(
                    id=str(uuid.uuid4()),
                    content=request.initial_prompt,
                    sender="user",
                    timestamp=datetime.now()
                ),
                ChatMessage(
                    id=str(uuid.uuid4()),
                    content=gen_response.explanation,
                    sender="ai",
                    timestamp=datetime.now()
                )
            ]
        else:
            # Create default files for the template
            if request.template == "react":
                initial_files = {
                    "App.js": ProjectFile(
                        name="App.js",
                        content="export default function App() {\n  return <div>Hello World</div>;\n}",
                        language="javascript"
                    ),
                    "index.js": ProjectFile(
                        name="index.js",
                        content="import React from 'react';\nimport ReactDOM from 'react-dom/client';\nimport App from './App';\n\nconst root = ReactDOM.createRoot(document.getElementById('root'));\nroot.render(<App />);",
                        language="javascript"
                    )
                }
        
        # Create project
        project_data = {
            "id": project_id,
            "title": request.title,
            "description": request.description,
            "template": request.template,
            "files": {name: file.dict() for name, file in initial_files.items()},
            "chat_history": [msg.dict() for msg in initial_chat],
            "user_clerk_id": request.user_clerk_id,
            "created_at": datetime.now(),
            "updated_at": datetime.now()
        }
        
        projects_db[project_id] = project_data
        
        return ProjectResponse(
            id=project_id,
            title=request.title,
            description=request.description,
            template=request.template,
            files=initial_files,
            chat_history=initial_chat,
            user_clerk_id=request.user_clerk_id,
            created_at=project_data["created_at"],
            updated_at=project_data["updated_at"]
        )
        
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/user/{user_clerk_id}", response_model=List[ProjectResponse])
async def get_user_projects(user_clerk_id: str):
    """Get all projects for a user"""
    try:
        user_projects = []
        for project_data in projects_db.values():
            if project_data["user_clerk_id"] == user_clerk_id:
                # Convert stored data back to response format
                files = {}
                for name, file_data in project_data["files"].items():
                    files[name] = ProjectFile(**file_data)
                
                chat_history = []
                for msg_data in project_data["chat_history"]:
                    chat_history.append(ChatMessage(**msg_data))
                
                user_projects.append(ProjectResponse(
                    id=project_data["id"],
                    title=project_data["title"],
                    description=project_data["description"],
                    template=project_data["template"],
                    files=files,
                    chat_history=chat_history,
                    user_clerk_id=project_data["user_clerk_id"],
                    created_at=project_data["created_at"],
                    updated_at=project_data["updated_at"]
                ))
        
        return user_projects
        
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/{project_id}", response_model=ProjectResponse)
async def get_project(project_id: str):
    """Get a specific project"""
    try:
        if project_id not in projects_db:
            raise HTTPException(status_code=404, detail="Project not found")
        
        project_data = projects_db[project_id]
        
        # Convert stored data back to response format
        files = {}
        for name, file_data in project_data["files"].items():
            files[name] = ProjectFile(**file_data)
        
        chat_history = []
        for msg_data in project_data["chat_history"]:
            chat_history.append(ChatMessage(**msg_data))
        
        return ProjectResponse(
            id=project_data["id"],
            title=project_data["title"],
            description=project_data["description"],
            template=project_data["template"],
            files=files,
            chat_history=chat_history,
            user_clerk_id=project_data["user_clerk_id"],
            created_at=project_data["created_at"],
            updated_at=project_data["updated_at"]
        )
        
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.put("/{project_id}/files", response_model=ProjectResponse)
async def update_project_files(project_id: str, request: ProjectUpdate):
    """Update project files"""
    try:
        if project_id not in projects_db:
            raise HTTPException(status_code=404, detail="Project not found")
        
        project_data = projects_db[project_id]
        
        # Update files
        for name, file in request.files.items():
            project_data["files"][name] = file.dict()
        
        project_data["updated_at"] = datetime.now()
        
        # Convert back to response format
        files = {}
        for name, file_data in project_data["files"].items():
            files[name] = ProjectFile(**file_data)
        
        chat_history = []
        for msg_data in project_data["chat_history"]:
            chat_history.append(ChatMessage(**msg_data))
        
        return ProjectResponse(
            id=project_data["id"],
            title=project_data["title"],
            description=project_data["description"],
            template=project_data["template"],
            files=files,
            chat_history=chat_history,
            user_clerk_id=project_data["user_clerk_id"],
            created_at=project_data["created_at"],
            updated_at=project_data["updated_at"]
        )
        
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/{project_id}/chat", response_model=ChatResponse)
async def chat_with_project(project_id: str, request: ChatRequest):
    """Chat about a project and potentially update files"""
    try:
        if project_id not in projects_db:
            raise HTTPException(status_code=404, detail="Project not found")
        
        project_data = projects_db[project_id]
        
        # Convert current files to ProjectFile objects
        current_files = {}
        for name, file_data in project_data["files"].items():
            current_files[name] = ProjectFile(**file_data)
        
        # Chat with AI
        chat_response = await project_service.chat_with_ai(request, current_files)
        
        # Add user message to chat history
        user_message = ChatMessage(
            id=str(uuid.uuid4()),
            content=request.message,
            sender="user",
            timestamp=datetime.now()
        )
        project_data["chat_history"].append(user_message.dict())
        
        # Add AI response to chat history
        ai_message = ChatMessage(
            id=str(uuid.uuid4()),
            content=chat_response.message,
            sender="ai",
            timestamp=chat_response.timestamp
        )
        project_data["chat_history"].append(ai_message.dict())
        
        # Update files if AI provided updates
        if chat_response.updated_files:
            for name, file in chat_response.updated_files.items():
                project_data["files"][name] = file.dict()
        
        project_data["updated_at"] = datetime.now()
        
        return chat_response
        
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.delete("/{project_id}")
async def delete_project(project_id: str):
    """Delete a project"""
    try:
        if project_id not in projects_db:
            raise HTTPException(status_code=404, detail="Project not found")
        
        del projects_db[project_id]
        return {"message": "Project deleted successfully"}
        
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))